### PR TITLE
tests: alter connection for zippy

### PR DESCRIPTION
**Pull request by nrainer-materialize**
This adds the `AlterSourceConnection` scenario to zippy.

Nightly: https://buildkite.com/materialize/nightlies/builds?branch=nrainer-materialize%3Atests%2Falter-connection-for-zippy

### CODE REVIEW

File: misc/python/materialize/zippy/source_actions.py
Comment (sploiselle) [inline misc/python/materialize/zippy/source_actions.py:115]: I fear that this doesn't test much as it is essentially a nop

Comment (nrainer-materialize) [reply misc/python/materialize/zippy/source_actions.py:115]: What exactly do you mean? That this statement as such is a NOP or that the test is a NOP?
If it is about the statement, then I will need to come up with something else. If it is about the test, I want to highlight that this is only a piece that is inserted into a sequence of actions.

Comment (sploiselle) [reply misc/python/materialize/zippy/source_actions.py:115]: This `ALTER` action sets the connection to the same value it already has--that's what I mean by NOP. It doesn't change any behavior--e.g. if this statement only parsed and returned an OK, this test would not uncover that.

Comment (nrainer-materialize) [reply misc/python/materialize/zippy/source_actions.py:115]: My understanding is that it will change the connection to `USING SSH TUNNEL thancred` or drop that from the connection definition. That is, it is supposed to change to the respective opposite (note the `not` in `if not self.source.uses_ssh_tunnel`).

Or do I miss something?

Comment (sploiselle) [reply misc/python/materialize/zippy/source_actions.py:115]: Oh, so sorry--I missed [this line](https://github.com/MaterializeInc/materialize/pull/23107/files#diff-4b4e49e5cd414d9531c8acbcdaab15ff12796c41a3684fabbc2e9ededf89a11eR121) when reading and didn't see it getting toggled back and forth!

Comment (nrainer-materialize) [reply misc/python/materialize/zippy/source_actions.py:115]: No, problem!

Comment (nrainer-materialize) [reply misc/python/materialize/zippy/source_actions.py:115]: I should maybe add a comment to make that more obvious.

File: misc/python/materialize/checks/mzcompose_actions.py
Comment (def-) [inline misc/python/materialize/checks/mzcompose_actions.py:72]: At this point we might also give it a more neutral/professional name :D

**Review by def- — APPROVED** [review]

File: misc/python/materialize/zippy/source_actions.py
Comment (philip-stoev) [inline misc/python/materialize/zippy/source_actions.py:60]: You need `SourceExists` as a requirement here as well.

**Review by philip-stoev — APPROVED** [review]
Thanks, this seems to be the right approach.

You mentioned that it does not work . What happens?

**Comment by nrainer-materialize** [comment]
> You mentioned that it does not work . What happens?

Some strange error without much details:
```
mzcompose: test case workflow-default failed: <class 'materialize.ui.UIError'>: running docker compose failed (exit status 1)
mzcompose: error: at least one test case failed
```

I will retry it in the scratch machine.

**Comment by nrainer-materialize** [comment]
> You mentioned that it does not work . What happens?

This happens without much details:
```
> ALTER SYSTEM SET enable_unmanaged_cluster_replicas = true;
> CREATE CLUSTER storage REPLICAS (r2 (
                STORAGECTL ADDRESSES ['storaged:2100'],
                STORAGE ADDRESSES ['storaged:2103'],
                COMPUTECTL ADDRESSES ['storaged:2101'],
                COMPUTE ADDRESSES ['storaged:2102'],
                WORKERS 4
            ))
[+] Stopping 1/1
 ✔ Container zippy-materialized-1  Stopped                                                                                                                                                0.1s 
Going to remove zippy-materialized-1
[+] Removing 1/0
 ✔ Container zippy-materialized-1  Removed                                                                                                                                                0.0s 
[+] Building 0.0s (0/0)                                                                                                                                                   docker:desktop-linux
[+] Running 1/1
 ✔ Container zippy-testdrive-1  Healthy                                                                                                                                                   0.0s 
service "materialized" is not running
sanity_restart disabled by override(), keeping it disabled
mzcompose: test case workflow-default failed: <class 'materialize.ui.UIError'>: running docker compose failed (exit status 1)
mzcompose: error: at least one test case failed
```

It also fails in scratch.
@philip-stoev, what do you suggest to get more insights / logs?

File: misc/python/materialize/zippy/source_actions.py
Comment (nrainer-materialize) [reply misc/python/materialize/zippy/source_actions.py]: ✅

**Comment by nrainer-materialize** [comment]
> It also fails in scratch. @philip-stoev, what do you suggest to get more insights / logs?

I guess I can use `bin/mzcompose --find zippy logs -f`.

There is a ton of errors like
```
zippy-schema-registry-1  | [2023-11-13 14:50:32,811] INFO 127.0.0.1 - - [13/Nov/2023:14:50:32 +0000] "GET / HTTP/1.1" 200 2 "-" "curl/7.61.1" GETsT (io.confluent.rest-utils.requests)
zippy-zookeeper-1        | [2023-11-13 14:50:33,865] WARN Unexpected exception (org.apache.zookeeper.server.NIOServerCnxn)
zippy-zookeeper-1        | EndOfStreamException: Unable to read additional data from client, it probably closed the socket: address = /127.0.0.1:53440, session = 0x0
zippy-zookeeper-1        | 	at org.apache.zookeeper.server.NIOServerCnxn.handleFailedRead(NIOServerCnxn.java:163)
zippy-zookeeper-1        | 	at org.apache.zookeeper.server.NIOServerCnxn.doIO(NIOServerCnxn.java:326)
zippy-zookeeper-1        | 	at org.apache.zookeeper.server.NIOServerCnxnFactory$IOWorkRequest.doWork(NIOServerCnxnFactory.java:522)
zippy-zookeeper-1        | 	at org.apache.zookeeper.server.WorkerService$ScheduledWorkRequest.run(WorkerService.java:154)
zippy-zookeeper-1        | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
zippy-zookeeper-1        | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
zippy-zookeeper-1        | 	at java.base/java.lang.Thread.run(Thread.java:829)
```

I will try to find other stuff in the crowded log. 😀

**Comment by sploiselle** [comment]
> service "materialized" is not running

IME this typically occurs because of the `sanity_restart` flag

**Comment by def-** [comment]
> sanity_restart disabled by override(), keeping it disabled

So shouldn't be related to sanity-restart. There could be a panic or a FATAL during startup.

File: test/zippy/mzcompose.py
Comment (def-) [inline test/zippy/mzcompose.py:187]: See line 183, materialized is removed there, so we can't set up the ssh connection here.

**Review by def- — CHANGES_REQUESTED** [review]
Ok, found the issue, it's the new setup_default_ssh_test_connection, which can only be run if `materialized` is running, but 2 lines above it it's removed.

File: test/zippy/mzcompose.py
Comment (nrainer-materialize) [reply test/zippy/mzcompose.py]: > See line 183, materialized is removed there, so we can't set up the ssh connection here.

Thank you @def-, this (plus upping `ssh-bastion-host`) was the fix.

**Comment by nrainer-materialize** [comment]
@sploiselle: Commits https://github.com/MaterializeInc/materialize/pull/23107/commits/76e9beaa40a5593eb72fb52186eee5999edf03f1 and https://github.com/MaterializeInc/materialize/pull/23107/commits/41cfddc38ce7c9fda7bac79efcbe6fc78a253986 are also ready to be cherry-picked. Although, there is a slight risk that the flakiness of "Zippy User Tables" is increased due to the inclusion of the ssh-bastion (see also https://github.com/MaterializeInc/materialize/issues/18628).

You can either cherry-pick these two commits onto your branch or I can merge this branch once yours is merged.

File: misc/python/materialize/zippy/source_actions.py
Comment (philip-stoev) [inline misc/python/materialize/zippy/source_actions.py:139]: AlterSourceConnection does not provide any additional capabilities, it merely operates on existing objects. So `provides()` needs to return an empty set (or not be defined at all, so the superclass provides() applies, which returns an empty set. This explains the "source specified more than once" error you are getting in the CI.

**Comment by nrainer-materialize** [comment]
[Nightly](https://buildkite.com/materialize/nightlies/builds/5355) is green now.

**Review by def- — APPROVED** [review]

**Pull request merged**
